src/main/java/org/radarbase/management/service/UserService.java

```diff
@@ -189,6 +204,45 @@ public User createUser(UserDTO userDto) {
                     };
                 })
                 .collect(Collectors.toSet());
+
+        checkAuthorityForRoleChange(roles, oldRoles);
+
+        return roles;
+    }
+
+    private void checkAuthorityForRoleChange(Set<Role> roles, Set<Role> oldRoles)
+            throws NotAuthorizedException {
+        var updatedRoles = new HashSet<>(roles);
+        updatedRoles.removeAll(oldRoles);
+        for (Role r : updatedRoles) {
+            checkAuthorityForRoleChange(r);
+        }
+
+        var removedRoles = new HashSet<>(oldRoles);
+        removedRoles.removeAll(roles);
+        for (Role r : removedRoles) {
+            checkAuthorityForRoleChange(r);
+        }
+    }
+
+    private void checkAuthorityForRoleChange(Role role)
+            throws NotAuthorizedException {
+        switch (role.getRole().scope()) {
+            case GLOBAL -> checkGlobalPermission(token, ROLE_UPDATE);
+            case ORGANIZATION -> checkPermissionOnOrganization(token, ROLE_UPDATE,
+                    role.getOrganization().getName());
+            case PROJECT -> {
+                if (role.getProject().getOrganization() != null) {
+                    checkPermissionOnOrganizationAndProject(token, ROLE_UPDATE,
+                            role.getProject().getOrganization().getName(),
+                            role.getProject().getProjectName());
+                } else {
+                    checkPermissionOnProject(token, ROLE_UPDATE,
+                            role.getProject().getProjectName());
+                }
+            }
+            default -> { }
```

Comment:
If we get an unexpected scope here, doesn't it indicate an exception-worthy bug?

Reply:
Yes... In fact I only added the default branch to silence a PMD warning -- it's already an exhaustive switch for all enum possibilities. I'll remove it and then it will be a compilation error instead if it's not exhaustive.

Reply:
Okay, if this is intentional I'm fine with it.

( However, a need to suppress a warning/ignore a newly introduced rule always leads to an obvious question :) )